pubnub-kotlin/src/main/kotlin/com/pubnub/internal/PubNubImpl.kt

```diff
@@ -0,0 +1,1442 @@
+     * @param inputString String to be decrypted.
+     *
+     * @return String containing the decryption of `inputString` using `cipherKey`.
+     * @throws PubNubException throws exception in case of failed decryption.
+     */
+    override fun decrypt(inputString: String): String = corePubNubClient.decrypt(inputString)
+
+    /**
+     * Perform Cryptographic decryption of an input string using a cipher key.
+     *
+     * @param inputString String to be decrypted.
+     * @param cipherKey cipher key to be used for decryption. Default is [PNConfiguration.cipherKey]
+     *
+     * @return String containing the decryption of `inputString` using `cipherKey`.
+     * @throws PubNubException throws exception in case of failed decryption.
+     */
+    override fun decrypt(
+        inputString: String,
+        cipherKey: String?,
+    ): String = corePubNubClient.decrypt(inputString, cipherKey)
+
+    /**
+     * Perform Cryptographic decryption of an input stream using provided cipher key.
+     *
+     * @param inputStream InputStream to be encrypted.
+     * @param cipherKey Cipher key to be used for decryption. If not provided [PNConfiguration.cipherKey] is used.
+     *
+     * @return InputStream containing the encryption of `inputStream` using `cipherKey`.
+     * @throws PubNubException Throws exception in case of failed decryption.
+     */
+    override fun decryptInputStream(
+        inputStream: InputStream,
+        cipherKey: String?,
+    ): InputStream = corePubNubClient.decryptInputStream(inputStream, cipherKey)
+
+    /**
+     * Perform Cryptographic encryption of an input string and a cipher key.
+     *
+     * @param inputString String to be encrypted.
+     * @param cipherKey Cipher key to be used for encryption. Default is [PNConfiguration.cipherKey]
+     *
+     * @return String containing the encryption of `inputString` using `cipherKey`.
+     * @throws PubNubException Throws exception in case of failed encryption.
+     */
+    override fun encrypt(
+        inputString: String,
+        cipherKey: String?,
+    ): String = corePubNubClient.encrypt(inputString, cipherKey)
+
+    /**
+     * Perform Cryptographic encryption of an input stream using provided cipher key.
+     *
+     * @param inputStream InputStream to be encrypted.
+     * @param cipherKey Cipher key to be used for encryption. If not provided [PNConfiguration.cipherKey] is used.
+     *
+     * @return InputStream containing the encryption of `inputStream` using `cipherKey`.
+     * @throws PubNubException Throws exception in case of failed encryption.
+     */
+    override fun encryptInputStream(
+        inputStream: InputStream,
+        cipherKey: String?,
+    ): InputStream = corePubNubClient.encryptInputStream(inputStream, cipherKey)
+
+    /**
+     * Force the SDK to try and reach out PubNub. Monitor the results in [SubscribeCallback.status]
+     */
+    override fun reconnect(timetoken: Long) = corePubNubClient.reconnect(timetoken)
```

Comment:
Why we need to override reconnect method here? In BasePubNubImpl there is reconnect method that seems to do the same job.

Reply:
you're correct, I moved some more methods from PubNubImpl to BasePubNubImpl instead, including unsubscribeAll, destroy, parseToken, setToken. If you notice any more methods that can be "shared" between clients and moved to BasePubNubImpl let me know!